File: GFG POTD/022-September-2025/003-reverse-a-doubly-linked-list.java
// https://www.geeksforgeeks.org/problems/reverse-a-doubly-linked-list/1




class Node {
    int data;
    Node next;
    Node prev;

    Node(int data) {
        this.data = data;
        this.next = null;
        this.prev = null;
    }
}

class Solution {
    // TC : O(N)
    // SC : O(1)
    public Node reverse(Node head) {
        Node prevN = null;
        Node curr  = head;
        
        while (curr != null) {
            Node forw = curr.next;
            curr.next = prevN;
            curr.prev = forw;
            
            prevN = curr;
            curr  = forw;
        }
        
        head = prevN;
        return head;
    }
}

File: GFG POTD/022-September-2025/003-reverse-a-doubly-linked-list.cpp
// https://www.geeksforgeeks.org/problems/reverse-a-doubly-linked-list/1/




#include <bits/stdc++.h>
using namespace std;


struct Node {
    int data;
    Node* next;
    Node* prev;
    Node (int x) {
        data = x;
        next = NULL;
        prev = NULL;
    } 
};

class Solution{
    public:
    // TC : O(N)
    // SC : O(1)
    Node* reverse(Node *head) {
        Node* prevN = NULL;
        Node* curr = head;
        
        while (curr) {
            Node* nextN = curr -> next;
            curr -> next = prevN;
            curr -> prev = nextN;
            
            // update the prev and curr
            prevN = curr;
            curr = nextN;
        }
        
        // update the head
        head = prevN;
        
        return head;
    }
};


